File: 1285-balance-a-binary-search-tree/balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* balanceBST(TreeNode* root) {
        if (root == nullptr) return nullptr;

        TreeNode* vineHead = new TreeNode(0);
        vineHead->right = root;
        
        TreeNode* current = vineHead;

        // Create backbone by right rotations
        while (current->right != nullptr) {
            if (current->right->left != nullptr) {
                rightRotate(current, current->right);
            }
            else {
                current = current->right;
            }
        }
        // Count nodes
        int count=0;
        current = vineHead->right;
        while (current!=nullptr) {
            count+=1;
            current = current->right;
        }
        
        int m = pow(2, floor(log2(count+1))) -1;

        // Make (count - m) left rotations to partially balance the tree
        doRotations(vineHead, count-m);
        // Further perform left rotations until m<=1 to balance the BST
        while(m>1) {
            m /= 2;
            doRotations(vineHead, m);
        }

        TreeNode* res = vineHead->right;
        delete vineHead;
        return res;

    }
private:
    void rightRotate(TreeNode* parent, TreeNode* node) {
        TreeNode* temp = node->left;
        node->left = temp->right;
        temp->right = node;
        parent->right = temp;
    }

    void leftRotate(TreeNode* parent, TreeNode* node) {
        TreeNode* temp = node->right;
        node->right = temp->left;
        temp->left = node;
        parent->right = temp;
    }

    void doRotations(TreeNode* vineHead, int count) {
        TreeNode* leftR = vineHead;
        for (int i=0; i<count; ++i) {
            TreeNode* tmp = leftR->right;
            leftRotate(leftR, tmp);
            leftR = leftR->right;
        }
    }
};

File: 1285-balance-a-binary-search-tree/balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:

        if not root:
            return None

        inord = []

        
        def inorder(root):
            if root is None:
                return
            
            if root.left:
                inorder(root.left)

            inord.append(root.val)

            if root.right:
                inorder(root.right)
            
        inorder(root)
        # print(inord)
        if len(inord) <= 1:
            return root

        new_root = TreeNode(0)
        q = [(new_root, 0, len(inord)-1 )]

        while q:
            node, l, r = q.pop(0)
            mid = l +(r - l) // 2
            node.val = inord[mid]
            
            if l <= mid-1:
                node.left = TreeNode(0)
                q.append((node.left, l, mid-1))
            if mid+1 <= r:
                node.right = TreeNode(0)
                q.append((node.right, mid+1, r))

            

        return new_root




        